geo: change encoding of geo types to a flat buffer

otan opened the pull request:
This PR changes the SpatialObject type to use the more efficient geom.T
method of storing a flat coord buffer, and ends/endss for multi-shaped
lines. Parsing EWKB to GeomT is expensive (as it is in GEOS as well), so
avoiding it where possible is ideal.

* Changed Shape to infer an actual Shape, and ShapeType to be the Shape
  enum we have today.
* Implemented various SpatialObject <-> GeomT transformations.
* Change `EWKBTo*` methods to be named `SpatialObjectTo`, avoiding a
  serialization step.
* Fix various compilation issues and bugs arising from invalid cloning
  for ShortestLine/LongestLine functions.
* Fixed Summary function to be more in line with PostGIS.
* Changed naming in types and sql.y package to reflect new reality.

Things I have omitted for the sake of commit brevity:
* GEOS to use GEOSSequence instead of EWKB for sending around.
* Various functions which involve heavy refactors (EWKB, EWKBHex).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50634)
<!-- Reviewable:end -->

otan committed:
geo: change encoding of geo types to a flat buffer

This PR changes the SpatialObject type to use the more efficient geom.T
method of storing a flat coord buffer, and ends/endss for multi-shaped
lines. Parsing EWKB to GeomT is expensive (as it is in GEOS as well), so
avoiding it where possible is ideal.

* Changed Shape to infer an actual Shape, and ShapeType to be the Shape
  enum we have today.
* Implemented various SpatialObject <-> GeomT transformations.
* Change `EWKBTo*` methods to be named `SpatialObjectTo`, avoiding a
  serialization step.
* Fix various compilation issues and bugs arising from invalid cloning
  for ShortestLine/LongestLine functions.
* Fixed Summary function to be more in line with PostGIS.
* Changed naming in types and sql.y package to reflect new reality.

Things I have omitted for the sake of commit brevity:
* GEOS to use GEOSSequence instead of EWKB for sending around.
* Various functions which involve heavy refactors (EWKB, EWKBHex).

Release note: None

otan commented:
let me come back to this. this was not as effective as i thought it was.

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50634)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geopb/geopb.pb.go, line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50634#-MAw1o2I45EdYYcOIDbq:-MAw1o2J7Bhd0Sacxgsf:bmu4c89) ([raw file](https://github.com/cockroachdb/cockroach/blob/806d4cd3e4ed8dab4ef787d4ba7ed808fec9d106/pkg/geo/geopb/geopb.pb.go#L76)):*
> ```Go
> 
> // Shape represents a single shape.
> type Shape struct {
> ```

what is the plan wrt parameter passing to our GEOS functions? Will we construct a struct and pass it and then convert to `geos::geom::Geometry` in c++ before calling the GEOS functions?


<!-- Sent from Reviewable.io -->
